docs/zed/design.md

```diff
@@ -0,0 +1,483 @@
+  <pool-id-1>/
+    branches/
+      HEAD
+      TAIL
+      1.zng
+      2.zng
+      ...
+    commits/
+      <id1>.zng
+      <id2>.zng
+      ...
+    data/
+      <id1>.{zng,zst}
+      <id2>.{zng,zst}
+      ...
+    index/
+      <id1>-<index-id-1>.zng
+      <id1>-<index-id-2>.zng
+      ...
+      <id2>-<index-id-1>.zng
+      ...
+  <pool-id-2>/
+  ...
+```
+
+## Continuous Ingest
+
+While the description above is very batch oriented, the Zed lake design is
+intended to perform scalably for continuous streaming applications.  In this
+approach, many small commits may be continuously executed as data arrives and
+after each commit, the data is immediately readable.
+
+To handle this use case, the _journal_ of branch commits is designed
+to scale to arbitrarily large footprints as described earlier.
+
+## Derived Analytics
+
+To improve the performance of predictable workloads, many use cases of a
+Zed lake pre-compute _derived analytics_ or a particular set of _partial
+aggregations_.
+
+For example, the Brim app displays a histogram of event counts grouped by
+a category over time.  The partial aggregation for such a computation can be
+configured to run automatically and store the result in a pool designed to
+hold such results.  Then, when a scan is run, the Zed analytics engine
+recognizes when the DAG of a query can be rewritten to assemble the
+partial results instead of deriving the answers from scratch.
+
+When and how such partial aggregations are performed is simply a matter of
+writing Zed queries that take the raw data and produce the derived analytics
+while conforming to a naming model that allows the Zed lake to recognize
+the relationship between the raw data and the derived data.
+
+> TBD: Work out these details which are reminiscent of the analytics cache
+> developed in our earlier prototype.
+
+## Keyless Data
+
+This is TBD.  Data without a key should be accepted some way or another.
+One approach is to simply assign the "zero-value" as the pool key; another
+is to use a configured default value.  This would make key-based
+retention policies more complicated.
+
+Another approach would be to create a sub-pool on demand when the first
+keyless data is encountered, e.g., `pool-name.$nokey` where the pool key
+is configured to be "this".  This way, an app or user could query this pool
+by this name to scan keyless data.
+
+## Relational Model
+
+Since a Zed lake can provide strong consistency, workflows that manipulate
+data in a lake can utilize a model where updates are made to the data
+in place.  Such updates involve creating new commits from the old data
+where the new data is a modified form of the old data.  This provides
+emulation of row-based updates and deletes.
+
+If the pool-key is chosen to be "this" for such a use case, then unique
+rows can be maintained by trivially detected duplicates (because any
+duplicate row will be adjacent when sorted by "this") so that duplicates are
+trivially detected.
```

Comment:
I'm not sure quite what adjustment to suggest here, but I just wanted to point out that it currently appears tautological as written.